### PR TITLE
gfx: swapchain refactor

### PR DESCRIPTION
**Pull request by YunHsiao**
Swapchain as a first-class gfx concept. Multiple swapchain applications are finally possible.

https://github.com/cocos-creator/3d-tasks/issues/7899
https://github.com/cocos-creator/3d-tasks/issues/8156
https://github.com/cocos-creator/3d-tasks/issues/8167
https://github.com/cocos-creator/3d-tasks/issues/8476
https://github.com/cocos-creator/3d-tasks/issues/8481
https://github.com/cocos-creator/3d-tasks/issues/8496
https://github.com/cocos-creator/3d-tasks/issues/8798
https://github.com/cocos-creator/3d-tasks/issues/8876

### CODE REVIEW

**Comment by YunHsiao** [comment]
https://github.com/cocos-creator/3d-tasks/issues/7899

File: cocos/renderer/gfx-agent/SwapchainAgent.h [hunk 1]
```diff
@@ -26,19 +26,26 @@
...
+class CC_DLL SwapchainAgent final : public Agent<Swapchain> {
 public:
-    explicit SamplerValidator(Sampler *actor);
-    ~SamplerValidator() override;
+    explicit SwapchainAgent(Swapchain *actor);
+    ~SwapchainAgent() override;
+
+    uint32_t getWidth() const override { return _actor->getWidth(); }
+    uint32_t getHeight() const override { return _actor->getHeight(); }
+    SurfaceTransform getSurfaceTransform() const override { return _actor->getSurfaceTransform(); }
```
Comment (minggo) [inline cocos/renderer/gfx-agent/SwapchainAgent.h:41]: 加 inline?

File: cocos/renderer/gfx-base/GFXInputAssembler.h [hunk 2]
```diff
@@ -68,17 +61,13 @@ class CC_DLL InputAssembler : public GFXObject {
...
-    uint          _firstIndex     = 0;
-    uint          _vertexOffset   = 0;
-    uint          _instanceCount  = 0;
-    uint          _firstInstance  = 0;
-    uint          _attributesHash = 0;
+    uint32_t      _attributesHash = 0;
+
+    BufferList _vertexBuffers;
+    Buffer *   _indexBuffer    = nullptr;
+    Buffer *   _indirectBuffer = nullptr;
```
Comment (minggo) [inline cocos/renderer/gfx-base/GFXInputAssembler.h:68]: 统一用 {} 初始化？

File: cocos/renderer/gfx-base/GFXSampler.h [hunk 3]
```diff
@@ -1,69 +0,0 @@
...
-****************************************************************************/
-
-#pragma once
-
-#include "GFXObject.h"
-
-namespace cc {
-namespace gfx {
-
-class CC_DLL Sampler : public GFXObject {
```
Comment (minggo) [inline cocos/renderer/gfx-base/GFXSampler.h:33]: 为什么 Sampler 也不需要了？

Comment (minggo) [reply cocos/renderer/gfx-base/GFXSampler.h:33]: 看了，挪位置了。

**Review by minggo — APPROVED** [review]

File: cocos/renderer/gfx-agent/SwapchainAgent.h (hunk 1 above)
Comment (YunHsiao) [reply cocos/renderer/gfx-agent/SwapchainAgent.h:41]: 这些是 override，其实没法 inline

Comment (minggo) [reply cocos/renderer/gfx-agent/SwapchainAgent.h:41]: 了解

**Pull request merged**